NeoML/include/NeoML/TraditionalML/BytePairEncoder.h

```diff
@@ -0,0 +1,174 @@
+/* Copyright � 2017-2022 ABBYY Production LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <NeoML/NeoMLDefs.h>
+#include <NeoML/TraditionalML/WordDictionary.h>
+
+namespace NeoML {
+
+// Class that calculates byte-pair-encoding tokens.
+class NEOML_API CBpeIterativeTrainer {
+public:
+	CBpeIterativeTrainer();
+	CBpeIterativeTrainer( const CBpeIterativeTrainer& other ) = delete;
+	CBpeIterativeTrainer& operator=( const CBpeIterativeTrainer& other ) = delete;
+
+	void Initialize( const CArray<CArray<CString>>& trainSplittedWords, 
+		const CArray<long long>& trainWordCounts, int totalIterationsCount );
+
+	// Performs the calculation of new BPE tokens.
+	CWordDictionary RunIterations( int iterationCount );
+	CWordDictionary RunTotalIterations();
+
+	// Returns the number of iterations completed (== number of calculated token).
+	int IterationsCompletedCount() const { return iterationsCompletedCount; }
+	// Returns the total number of iterations.
+	int TotalIterationsCount() const { return totalIterationsCount; }
+	// Returns true if no more iterations can be performed.
+	bool IsCompleted() const;
+
+	// Serialization to archive.
+	void Serialize( CArchive& archive );
+
+private:
+	// The total number of iterations.
+	// The size of the dictionary returned from RunTotalIterations cannot exceed this value.
+	int totalIterationsCount;
+	// The number of completed iterations.
+	int iterationsCompletedCount;
+
+	// The dictionary of pairs of neighbour tokens.
+	CWordDictionary pairDictionary;
+
+	CArray<CArray<CString>> trainWords;
+	CArray<long long> trainCounts;
+
+	// Map: pair of neighbour tokens -> set of ids of words containing this pair of tokens.
+	typedef CMap<CString, CHashTable<int>> CPairReverseIndex;
+	CPairReverseIndex reverseIndex;
+
+	int calcIterationsCount( int requestedIterationsCount ) const;
+	void buildPairDictionary( CWordDictionary& newTokens );
+	bool runSingleIteration( CWordDictionary& newTokens );
+};
+
+// Class that encodes a word using byte-pair-encoding.
+class NEOML_API CBytePairEncoder {

```

Comment:
We need interfaces:
- `ITextEncoder` with methods `Encode` `Decode` `Size` and (maybe) `Serialize` (or `ITextTokenizer` and methods `Tokenize` etc.)
- `ITextBpeEncoder : public ITextEncoder` with additional BPE-specific method

And we need some trainer class which returns those interfaces. For reference, see `IModel`, `IGradientBoostModel` and `CGradientBoost`.